File: 23.동적 계획법 1/11054.가장 긴 바이토닉 부분 수열/Main.java
import java.io.*;
import java.util.*;

public class Main {
    static int findReplaceIndex(Stack<Integer> stack, int num) {
        int l = 0, r = stack.size() - 1;

        while (l < r) {
            int mid = l + (r - l) / 2;

            if (stack.get(mid) >= num) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return r;
    }

    static int findReplaceIndex2(Stack<Integer> stack, int num) {
        int l = 0, r = stack.size() - 1;

        while (l < r) {
            int mid = l + (r - l) / 2;

            if (stack.get(mid) <= num) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return r;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N + 1];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int result = 0;
        Stack<Integer> stack = new Stack<>();
        ;
        for (int c = 0; c < N; c++) {

            int leftCount = 0;
            stack.clear();
            for (int i = 0; i < c; i++) {
                if (arr[i] >= arr[c])
                    continue;
                if (stack.isEmpty() || stack.peek() < arr[i]) {
                    stack.push(arr[i]);
                    leftCount++;
                    continue;
                }
                stack.set(findReplaceIndex(stack, arr[i]), arr[i]);
            }

            int rightCount = 0;
            stack.clear();
            for (int i = c + 1; i < N; i++) {
                if (arr[i] >= arr[c])
                    continue;
                if (stack.isEmpty() || stack.peek() > arr[i]) {
                    stack.push(arr[i]);
                    rightCount++;
                    continue;
                }
                stack.set(findReplaceIndex2(stack, arr[i]), arr[i]);
            }

            result = Math.max(result, leftCount + rightCount + 1);
        }

        bw.write(String.valueOf(result));
        bw.close();
        br.close();
    }
}

File: 23.동적 계획법 1/11054.가장 긴 바이토닉 부분 수열/main.py
import sys

input = sys.stdin.readline


def find_replace_index(lst, num):
    l, r = 0, len(lst) - 1
    while l < r:
        mid = l + (r - l) // 2
        if lst[mid] >= num:
            r = mid
        else:
            l = mid + 1
    return r


def find_replace_index2(lst, num):
    l, r = 0, len(lst) - 1
    while l < r:
        mid = l + (r - l) // 2
        if lst[mid] <= num:
            r = mid
        else:
            l = mid + 1
    return r


if __name__ == "__main__":
    N = int(input())
    arr = list(map(int, input().split()))

    result = 0

    for center in range(0, N):
        center_num = arr[center]

        left_count = 0
        stack = []
        for i in range(0, center):
            # 왼쪽 부분 구하기
            if arr[i] >= center_num:
                continue
            if not stack or stack[-1] < arr[i]:
                stack.append(arr[i])
                left_count += 1
                continue
            stack[find_replace_index(stack, arr[i])] = arr[i]

        right_count = 0
        stack2 = []
        for i in range(center + 1, N):
            # 오른쪽 부분 구하기
            if arr[i] >= center_num:
                continue
            if not stack2 or stack2[-1] > arr[i]:
                stack2.append(arr[i])
                right_count += 1
                continue
            stack2[find_replace_index2(stack2, arr[i])] = arr[i]

        result = max(result, left_count + right_count + 1)
    print(result)
